### PR TITLE
[rai] changed the extractor for the new site and dismisison of the older one

### PR DESCRIPTION
**Pull request by timendum**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Rai recently changed its streaming website to raiplay.it and axed the old one, I heavy changed the extractor to reflect the new changes.

### CODE REVIEW

**Comment by timendum** [comment]
Can I help merging this issue in any way?

File: youtube_dl/extractor/rai.py [hunk 1]
```diff
@@ -55,7 +55,118 @@ def _extract_relinker_formats(self, relinker_url, video_id):
...
+        if 'images' in media:
+            for key in media.get('images'):
+                if media.get('images').get(key):
+                    thumbnails.append({
+                        'url': media.get('images').get(key).replace('[RESOLUTION]', self._RESOLUTION)
+                    })
+
+        formats = []
+        duration = None
+        if 'video' in media:
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: If `video` not in `media` empty formats will be returned that does not make any sense.

File: youtube_dl/extractor/rai.py [hunk 2]
```diff
@@ -55,7 +55,118 @@ def _extract_relinker_formats(self, relinker_url, video_id):
...
+
+        canonical_url = self._og_search_url(webpage)
+
+        media = self._download_json('%s?json' % canonical_url,
+                                    video_id, 'Downloading video JSON')
+
+        thumbnails = []
+        if 'images' in media:
+            for key in media.get('images'):
+                if media.get('images').get(key):
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: `for key, value in media.get('images', {}).items():`

File: youtube_dl/extractor/rai.py [hunk 3]
```diff
@@ -55,7 +55,118 @@ def _extract_relinker_formats(self, relinker_url, video_id):
...
+            'title': self._og_search_title(webpage),
+            'description': self._og_search_description(webpage),
+            'uploader': media.get('channel'),
+            'duration': duration,
+            'thumbnails': thumbnails,
+            'formats': formats
+        }
+
+class RaiItIE(RaiBaseIE):
+    _VALID_URL = r'https?://.+\.(rai|rainews)\.it/dl/.+?-(?P<id>[\da-f]{8}-[\da-f]{4}-[\da-f]{4}-[\da-f]{4}-[\da-f]{12})\.html'
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: Don't capture groups you don't use.

File: youtube_dl/extractor/rai.py [hunk 4]
```diff
@@ -55,7 +55,118 @@ def _extract_relinker_formats(self, relinker_url, video_id):
...
+        return {
+            'id': video_id,
+            'title': self._og_search_title(webpage),
+            'description': self._og_search_description(webpage),
+            'uploader': media.get('channel'),
+            'duration': duration,
+            'thumbnails': thumbnails,
+            'formats': formats
+        }
+
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: Read and follow code conventions. Check code with flake8.

File: youtube_dl/extractor/rai.py [hunk 5]
```diff
@@ -55,7 +55,118 @@ def _extract_relinker_formats(self, relinker_url, video_id):
...
+class RaiItIE(RaiBaseIE):
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: `RaiIE`.

File: youtube_dl/extractor/rai.py [hunk 6]
```diff
@@ -1,17 +1,17 @@
...
     determine_ext,
-    ExtractorError,
     find_xpath_attr,
     fix_xml_ampersands,
     int_or_none,
     parse_duration,
     unified_strdate,
     update_url_query,
-    xpath_text,
+    xpath_text
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: `,` is fine here.

File: youtube_dl/extractor/rai.py [hunk 7]
```diff
@@ -55,7 +55,118 @@ def _extract_relinker_formats(self, relinker_url, video_id):
...
+        }
+    }]
+    _RESOLUTION = '600x400'
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        webpage = self._download_webpage(url, video_id)
+
+        canonical_url = self._og_search_url(webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/rai.py]: What's the point of this? `og:url` is always the same as the original `url`. If not provide test for that.

**Comment by timendum** [comment]
Thanks for the review, I've fixed all the points.

About the two comments are not outdated: the first one is fixed by adding a new blank line, the other by adding a new test case.

**Comment by 9peppe** [comment]
Guys, it's been one and a half months, and the extractor apparently works for me.

Is there anything I can do to help this get merged?

**Comment by rabblac** [comment]
Unfortunately some tests are failing. In particular:
- on RaiPlayIE, title is the right one only when `source` parameter is specified in the URL, otherwise it will have ` - video - RaiPlay` appended. Can't suggest a fix since I don't know project recommendations.
- on RaiIE, downloaded files have mp4 extension instead of flv. Easy fix: [rai.diff.txt](https://github.com/rg3/youtube-dl/files/714262/rai.diff.txt)

**Pull request closed without merging**